### ISSUE TITLE
Issue Prioritization as of 11/2021

### ISSUE THREAD

**Issue by Kethku** [neovide/neovide#1061]
@dineshKumar777 came up with a list of issues categorized by priority which follow slightly edited. I have also split them into predicted easy and hard by my estimation of how much work would be required.

This is just a prioritization of issues that are worth addressing in the relatively short term. This does not mean things not on this list will be left forever, just that these are high profile issues that are worth focusing on for now from a planning point of view.

## High Priority Easy
- [x] #1004 
  - This has been working in the past so I think something silly has happened in some PR which got merged incorrectly. I don't expect this to be very hard.
- [x] #501 
  - This shouldn't be too complicated. I suspect this is just something specific to the way we report keybindings or something similar. If not then this is likely some issue upstream that we just need to report. Just needs investigation
- [x] WSL Related issues
  - I believe these are mostly due to a lack of debugging or visible errors. I have yet to encounter a wsl problem that couldn't be solved with some fiddling. Its just theres more moving parts than expected
- [x] #1057 
  - This is a common source of friction for people trying to use neovide. Its not obvious what names should be specified in order to get the font to be picked up properly by the rendering system. If we could provide a way to have auto complete for the guifont option, that would go a long way toward making this easier for the user. Maybe we could expose a list on startup in some global variable which has all the font names known to the system. Then the user could look through for the one they want and be sure to get it from the start rather than the current trial and error.
- [x] #1091 
  - Neovide has extensive hacks in place to work around a series of bugs in neovim 0.5 with regards to mouse positioning when multigrid is enabled. I believe most of the issues have been fixed in 0.6, so we need to update the readme to specify a new minimum version, and remove the hacks. Should mostly be deleting code. 

## High Priority Hard
- [x] #740 
  - This is hard to track down. There is a nebulous report that performance was worse with the winit move, but its not clear why or how to address the problem. I've attempted some profiling like i have done in the past but with only limited success. Some brainstorming for ideas might help here.
- [ ] #843  
  - This is hard because its not clear how to debug the problem. It happens infrequently and unreliably so its just sorta luck of the draw whether we find the issue. Maybe this would be improved by better error reporting?
- [ ] Better startup/error readouts
  - Currently errors often result in a silent crash without any information or avenue to figure out what went wrong. We should design a way to report better information and display it to the user rather than just bailing silently.
- [ ] #445 
  - Progress has been made toward better keyboard handling, but theres still a ton of work that is yet to merge upstream. Today neovide uses a fork that I constructed of winit which combines 3 outstanding PRs into something somewhat usable, but its still not there yet. This is complicated because not only is our progress slow, but upstream progress is also steady but very slow. Some eyes or just attention on the current status might help make the wait more bearable.
- [x] #589 
  - This is another hisenbug. Its not clear what causes this issue or why it appears sometimes but not others. Its also multi faceted; part of it might be the anti aliasing applied to the characters, part might be off by one errors in the font positioning, part might be due to how we save cpu cycles by not rerendering everything and instead copying the canvas when scrolling. Maybe its a combination of all of the above. Its not clear to me and will require more investigation
 
## Low Priority Easy
- [x] #887 
- [x] #1030 
- [x]  #966 
  - This shouldn't be too bad. We would just have to specify using skia (or swash) options what width we want. And then modify the guifont parser to parse the width flag and pass that along.
- [x] #1021 

## Low Priority Hard
- [ ] #1024 
  - Although not super difficult, this would require refactoring the font shaper to handle more than just a few fonts. I would argue this is closer to medium than hard but it is somewhat low priority because to my knowledge, other neovim editors do not support this. As such we will have to come up with a design for how these settings should be specified in the guifont variable as the public spec isn't clear about how one might configure it for this purpose.
- [ ] #768 
  - This is hard just because we need to be careful to pick and choose which ones really need implemented over the default behavior. I personally don't want to spend time and maintenance on features that don't provide a real benefit to the end user (nobody look at the cursor trails hahaha) but am absolutely willing to help somebody if they want to take it on themselves. My only hope is that there is a reason behind the design besides just esthetics and that it should feel at home with neovim.
- [ ] #720  #939 
  - EDIT: Due to some wonderful contributions, the relevant plugins have been made to work, so I have moved this problem to low priority hard. Very exciting :)
  - This is a gnarly one. Floating window transparency is specified by neovim as a buffer local option. Currently neovide only has infrastructure for syncing and reading global variables as settings. I think to fix this problem we will have to extend the setting system to read and keep synced with buffer local options. Further those buffer local options should be accessible by current buffer (or maybe via some kind of buffer id indexed dictionary). This is extra complicated because the settings system is backed by a very opaque procedural macro which is hard to change. *I have attempted fixing this myself 2 or 3 times and have failed to arrive at something workable* maybe somebody more brave or ambitious than I would have more luck.

**Comment by Kethku** [comment]
As of right now, the biggest issue blocking a new release is the intermittent crash that some folks (all??) are getting. Beyond that, the others are either workable or can be fixed in a subsequent release. Getting close I think.

**Comment by fredizzimo** [comment]
One common cause of a crash/hang is fixed in https://github.com/neovide/winit/pull/4. I'm not sure if that's the one affecting other people or not, but I can't really say that I have got any intermittent crashes, since I made those changes.

I think I might have had a few freezes or hangs, but I think all of those has been caused by Neovim, not Neovide, and because of my own mistakes, not really bugs. And if there are some remaining Neovide crashes for me, then they are quite rare, since I'm using it full time at work.

**Comment by enderger** [comment]
I think another blocker is the suspend crash, especially on laptops.

**Comment by fredizzimo** [comment]
For suspend crash there's also an open pull request https://github.com/neovide/winit/pull/5

**Comment by Kethku** [comment]
Ok I have merged both. @fredizzimo you're a legend. I've sent an invite to the org in case you have any desire to merge further fixes. This is a huge help

**Comment by enderger** [comment]
> Ok I have merged both. @fredizzimo you're a legend. I've sent an invite to the org in case you have any desire to merge further fixes. This is a huge help

It fixed the crash, but now #1120 is an issue

**Comment by fredizzimo** [comment]
@Kethku, thank you very much for the invitation, but I think I prefer to stay as a contributor. I don't really have the time needed, since I have so many other projects, a busy work, and a family. I'm also really not that confident in my Rust programming skills, at least not to perform any code reviews. 

But I will continue to contribute, at least with issues that are bothering me personally, and perhaps with some other things that does not take too much of my time, like debugging, and finding the root cause of issues.

You have made a really great piece of software, and I'm already using it full time, so I'm happy to make it even better!

**Comment by Kethku** [comment]
Okeydokey! I sent it in case you were interested. Already your contributions have made a big difference :)

**Comment by fredizzimo** [comment]
For the WSL issues. I think once this https://github.com/neovide/neovide/pull/1118 gets confirmed working for other people and merged, the only remaining one is to be able to specify the distribution to use https://github.com/neovide/neovide/issues/531. 

#1118 also improves the error messages when Neovim is not found inside WSL, so I think that should cover the error reporting.  I think all other error reporting should be the same as regular launching.

Copy/pasting reported in https://github.com/neovide/neovide/issues/914 works fine, using the normal nvim recommendation of using `win32yank` https://github.com/neovim/neovim/wiki/FAQ#how-to-use-the-windows-clipboard-from-wsl. Through WSLg, it's slightly more complicated, since Nvim autodetects xclip, but I have this in my `init.vim` file to fix that
```
let g:clipboard = {
          \   'name': 'win32yank-wsl',
          \   'copy': {
          \      '+': 'win32yank.exe -i --crlf',
          \      '*': 'win32yank.exe -i --crlf',
          \    },
          \   'paste': {
          \      '+': 'win32yank.exe -o --lf',
          \      '*': 'win32yank.exe -o --lf',
          \   },
          \   'cache_enabled': 0,
          \ }
```
But of course native copy paste would be nicer, and for TCP remote instances basically a requirement.

**Comment by shaunsingh** [comment]
Would like to nominate https://github.com/neovide/neovide/issues/447 for either High Priority Easy or Hard. As mentioned in this comment: https://github.com/neovide/neovide/issues/447#issuecomment-941597349, neovide fails to detect executables on path correctly, the offending code should be https://github.com/neovide/neovide/blob/9cfac28ac8add0e03e2cab46a5be034c77a78a9d/src/main.rs#L198-L214

> // printenv is the proper way to print env variables. using echo $PATH would break Fish.

Just wondering, the login shelll on macOS is always ZSH so is this an issue?

**Comment by Kethku** [comment]
I have no idea. @shaunsingh I am more than happy to defer to your expertise in this matter. If you think it would be better with echo path, that sounds reasonable to me

**Comment by shaunsingh** [comment]
I haven't dealt with this much before. I'll try playing around with it.

**Comment by Kethku** [comment]
Ok I think it's time to cut a new prerelease. There are a couple of high-profile bugs that are close but delaying longer isn't necessarily helpful. Unless anybody has any reason to wait, I'm going to make a new release tonight or tomorrow

**Comment by fredizzimo** [comment]
I think there are two quite bad issues left. the first one is the Mac issue mentioned above. https://github.com/neovide/neovide/issues/447. I made a suggestion at the bottom of the thread, but I have no experience with Mac or even a Mac to test on, so I don't know if it can be fixed that way. Otherwise I guess that we have to add to the release notes that the Mac version only works when launched through the terminal, at least that's how I understand it works right now.

The other one, is about snaps, they don't work on wayland https://github.com/neovide/neovide/issues/1060 and the GPU support is not setup correctly https://github.com/neovide/neovide/issues/1081. I have added comments to both of those, But I think someone with more experience need to make the actual changes, especially since the workflow builder is also failing due to the snaps currently. If these don't get fixed, I don't think we should even try to publish a Snap version.

**Comment by Kethku** [comment]
Those are fair points. So in your estimation should we block on those?

**Comment by shaunsingh** [comment]
I agree with @fredizzimo on blocking on those two issues.

**Comment by fredizzimo** [comment]
But issues seems quite easily solvable if we put some work into it, so yes, I think we can wait a few days and try to solve them. But obviously we shouldn't wait for too long, if we can't resolve them, since the Mac version still works from the terminal, and on Linux, you just have to build from source or use a package manager instead of snaps.

It would be nice if all platforms worked properly.

**Comment by Kethku** [comment]
Okeydokey! Better to land well. (I'm on vaca this weekend, but none of these things are blocked on me I think. Yall should merge things if you think its ready or fixes are done!)

**Comment by Kethku** [comment]
I'm pushing on the macos fix now, and have fixed a few other outstanding issues that were bothering me.

So I'm wondering if snap should still block a release. https://github.com/neovide/neovide/issues/1149 the convo here seems to be moving forward, but theres a number of issues, and honestly I have very little experience with linux, and even less with snap packages.

Are snap packages the main/only way for installing on linux? Would it be reasonable to give manual install instructions for now and call the release a beta? Should still delay the release until we have some more understanding of the problem space?

**Comment by fredizzimo** [comment]
I think `cargo install --git url --tag tag` or with the repository cloned `cargo install --path .` are decent alternatives. But the users most likely need to add `~/.cargo/bin` to the path manually.

I think we can do a minor release later that adds snap support.

On Arch and derivatives like Manjaro, we also have AUR support, so that's the preferred way on those distributions. The main package only needs to be flagged out of date when the version is released, and the maintainer should update it.

There are of course a lot of other Linux package managers we could add support for.

There are general ones like AppImage and Flatpack, just like Ubuntu is moving towards snaps, some other distributions are moving towards those. Then there's Nix (which also would add support for NixOS), Guix is very similar to that.

For Ubuntu, Debian and other derivatives like Mint, PopOS and MX, we could host our own PPA, or provide a dpkg, this covers a huge chunk of the Linux userbase.

For Fedora and derivatives there are RPMs. At least Fedora will most likely adopt it almost immediately if a package request was made. But I think we  need to do provide the actual package ourselves https://docs.fedoraproject.org/en-US/package-maintainers/Package_Review_Process/

Gentoo is also very popular, and they have ebuild packages https://wiki.gentoo.org/wiki/Submitting_ebuilds, and the process is similar to Fedora

I'm not sure how much of this we need to do by ourselves, and what will eventually get picked up by Linux distributions if the project gets popular. I only use Arch myself, and just a little bit of Ubuntu, so I don't know any details of other distributions.

For Windows, we should also update the scoop package that is already there https://github.com/neovide/neovide/issues/1109 . That should just be a matter of updating the file and send a pull request, like was initially done here https://github.com/ScoopInstaller/Extras/pull/3731

**Comment by shaunsingh** [comment]
> Are snap packages the main/only way for installing on linux? Would it be reasonable to give manual install instructions for now and call the release a beta? Should still delay the release until we have some more understanding of the problem space?

Its a good cross-distro way to handle it. Theres neovide packages for nix and the AUR iirc, but nothing for debian/apt based distros

**Comment by Kethku** [comment]
Okeydokey. New blockers cause delaying releases is fun :)

1. Environment variable for settings which can't be set in vim config is not usable on macos. So https://github.com/neovide/neovide/pull/1119 should be finished before release
2. Swap multigrid to be enabled by default. Upstream changes mean that our current mouse system works best when multigrid is enabled and I really don't want to special case it
3. Add settings for disabling smooth scrolling. For reasons I don't fully understand, it is jumpy but only sometimes. This is a main issue with multigrid today
4. Investigate weirdly offset floating windows (some plugins don't position windows correctly with multigrid enabled)
5. Startup issues in Macos (there is an open PR for this)

**Comment by Kethku** [comment]
Closing this as old
